src/trackedit/internal/au3/au3interaction.cpp

```diff
@@ -1303,6 +1305,82 @@ void Au3Interaction::moveTracksTo(const TrackIdList& trackIds, int to)
     projectHistory()->pushHistoryState("Move track", "Move track");
 }
 
+void Au3Interaction::insertSilence(const TrackIdList trackIds, secs_t begin, secs_t end, secs_t duration)
+{
+    auto prj = globalContext()->currentTrackeditProject();
+    auto& project = projectRef();
+
+    if (trackIds.empty()) {
+        auto& tracks = Au3TrackList::Get(project);
+        auto& trackFactory = ::WaveTrackFactory::Get(project);
+
+        sampleFormat defaultFormat = QualitySettings::SampleFormatChoice();
+        auto rate = ::ProjectRate::Get(project).GetRate();
+
+        auto track = trackFactory.Create(defaultFormat, rate);
+        track->SetName(tracks.MakeUniqueTrackName(Au3WaveTrack::GetDefaultAudioTrackNamePreference()));
+
+        track->InsertSilence(begin, duration);
+
+        tracks.Add(track);
+
+        trackedit::ITrackeditProjectPtr prj = globalContext()->currentTrackeditProject();
+        prj->notifyAboutTrackAdded(DomConverter::track(track.get()));
+        projectHistory()->pushHistoryState(muse::trc("trackedit", "Insert silence"), muse::trc("trackedit", "Insert silence"));
+        return;
+    }
+
+    for (const auto& trackId : trackIds) {
+        Au3WaveTrack* waveTrack = DomAccessor::findWaveTrack(project, Au3TrackId(trackId));
+
+        std::vector<Clip> clipsBefore;
```

Comment:
Couldn't `ITrackeditProject` have
```cpp
virtual std::vector<Clip> clipList(const TrackId& trackId) const = 0;
virtual muse::async::NotifyList<Clip> clipNotifyList(const TrackId& trackId) const = 0;
```
and then
```cpp
std::vector<au::trackedit::Clip> Au3TrackeditProject::clipList(const au::trackedit::TrackId& trackId) const
{
    const Au3WaveTrack* waveTrack = DomAccessor::findWaveTrack(*m_impl->prj, Au3TrackId(trackId));
    IF_ASSERT_FAILED(waveTrack) {
        return std::vector<au::trackedit::Clip>();
    }

    std::vector<au::trackedit::Clip> clips;
    for (const std::shared_ptr<const Au3WaveClip>& interval : waveTrack->Intervals()) {
        au::trackedit::Clip clip = DomConverter::clip(waveTrack, interval.get());
        clips.push_back(std::move(clip));
    }

    return clips;
}

muse::async::NotifyList<au::trackedit::Clip> Au3TrackeditProject::clipNotifyList(const au::trackedit::TrackId& trackId) const
{
    auto clips = clipList(trackId);
    async::ChangedNotifier<Clip>& notifier = m_clipsChanged[trackId];
    clips.setNotify(notifier.notify());

    return clips;
}
```
?

Reply:
@kryksyh what about this comment ?

Reply:
Replaced with track change notification.